internal/engine/upper.go

```diff
@@ -486,32 +495,66 @@ func handleStartProfilingAction(state *store.EngineState) {
 	state.IsProfiling = true
 }
 
-func handleFSEvent(
-	ctx context.Context,
-	state *store.EngineState,
-	event fswatch.TargetFilesChangedAction) {
+func handleFileWatchCreateEvent(_ context.Context, state *store.EngineState, action fswatch.FileWatchCreateAction) {
+	name := types.NamespacedName{Namespace: action.FileWatch.GetNamespace(), Name: action.FileWatch.GetName()}
+	state.FileWatches[name] = action.FileWatch
+}
 
-	if event.TargetID.Type == model.TargetTypeConfigs {
-		for _, f := range event.Files {
-			state.PendingConfigFileChanges[f] = event.Time
+func handleFileWatchUpdateEvent(_ context.Context, state *store.EngineState, action fswatch.FileWatchUpdateAction) {
+	name := types.NamespacedName{Namespace: action.FileWatch.GetNamespace(), Name: action.FileWatch.GetName()}
+	fw := state.FileWatches[name]
+	if fw == nil {
+		return
+	}
+	action.FileWatch.DeepCopyInto(fw)
+	processFileWatchStatus(state, name, &action.FileWatch.Status)
+}
+
+func handleFileWatchUpdateStatusEvent(_ context.Context, state *store.EngineState, action fswatch.FileWatchUpdateStatusAction) {
+	fw := state.FileWatches[action.Name]
+	if fw == nil {
+		return
+	}
+	action.Status.DeepCopyInto(&fw.Status)
+	processFileWatchStatus(state, action.Name, action.Status)
+}
+
+func processFileWatchStatus(state *store.EngineState, name types.NamespacedName, status *filewatches.FileWatchStatus) {
+	if status.Error != "" || len(status.FileEvents) == 0 {
+		return
+	}
+
+	// since the store is called on EVERY update, can always just look at the last event
+	latestEvent := status.FileEvents[len(status.FileEvents)-1]
+
+	// TODO(milas): should probably centralize this logic and ensure `:` can't exist in resource names
+	targetParts := strings.SplitN(name.Name, ":", 2)
```

Comment:
i wonder if these should be annotations on the FileWatch to know who should get triggered, at least to help with migration (rather than relying on the Name)

Reply:
Switched to a label (to mirror manifest name - I think this is correct since its immutable? but can easily switch to annotation) in 6c4ed98